File: app/src/main/java/com/example/punchpad2/NoteAdapter.java
package com.example.punchpad2;

import android.content.Context;
import android.graphics.Color;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class NoteAdapter extends RecyclerView.Adapter<NoteAdapter.NoteViewHolder> {

    private final List<NoteEntity> notes;
    private final Context context;
    private final boolean deleteMode;
    private final OnNoteChangedListener listener;

    public interface OnNoteChangedListener {
        void onNoteUpdated(NoteEntity note);
    }

    public NoteAdapter(Context context, List<NoteEntity> notes, boolean deleteMode, OnNoteChangedListener listener) {
        this.context = context;
        this.notes = notes;
        this.deleteMode = deleteMode;
        this.listener = listener;
    }

    @NonNull
    @Override
    public NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_note, parent, false);
        return new NoteViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteViewHolder holder, int position) {
        holder.bind(notes.get(position));
    }

    @Override
    public int getItemCount() {
        return notes.size();
    }

    class NoteViewHolder extends RecyclerView.ViewHolder {
        EditText noteText;
        ImageView starIcon;
        boolean isEditing = false;

        NoteViewHolder(View itemView) {
            super(itemView);
            noteText = itemView.findViewById(R.id.noteText);
            starIcon = itemView.findViewById(R.id.starIcon);
        }

        void bind(NoteEntity note) {
            noteText.setText(note.content);
            starIcon.setImageResource(note.isHidden ? R.drawable.ic_star_filled : R.drawable.ic_star_outline);

            noteText.setFocusable(false);
            noteText.setCursorVisible(false);
            noteText.setBackgroundColor(Color.TRANSPARENT);

            GestureDetector gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
                @Override
                public boolean onDoubleTap(MotionEvent e) {
                    note.isHidden = !note.isHidden;
                    starIcon.setImageResource(note.isHidden ? R.drawable.ic_star_filled : R.drawable.ic_star_outline);
                    Toast.makeText(context,
                            note.isHidden ? "Note hidden from preview." : "Note visible in preview.",
                            Toast.LENGTH_SHORT).show();
                    listener.onNoteUpdated(note);
                    return true;
                }

                @Override
                public boolean onSingleTapConfirmed(MotionEvent e) {
                    if (!isEditing) {
                        isEditing = true;
                        noteText.setFocusableInTouchMode(true);
                        noteText.setCursorVisible(true);
                        noteText.requestFocus();
                        noteText.setBackgroundColor(0x22000000);
                    }
                    return true;
                }
            });

            noteText.setOnTouchListener((v, event) -> gestureDetector.onTouchEvent(event));

            noteText.setOnFocusChangeListener((v, hasFocus) -> {
                if (!hasFocus) {
                    isEditing = false;
                    noteText.setCursorVisible(false);
                    noteText.setBackgroundColor(Color.TRANSPARENT);
                }
            });

            noteText.addTextChangedListener(new TextWatcher() {
                @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
                @Override public void afterTextChanged(Editable s) {}
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    note.content = s.toString();
                    listener.onNoteUpdated(note);
                }
            });

            if (deleteMode && note.isHidden) {
                starIcon.setOnClickListener(v ->
                        Toast.makeText(context, "Can't delete hidden notes.", Toast.LENGTH_SHORT).show());
            }
        }
    }
}

File: app/src/main/java/com/stratonotes/NoteAdapter.kt
package com.stratonotes

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.punchpad2.R

class NoteAdapter(
    private val context: Context,
    private val listener: OnNoteChangedListener
) : ListAdapter<NoteEntity, NoteAdapter.NoteViewHolder>(DIFF_CALLBACK) {

    interface OnNoteChangedListener {
        fun onNoteChanged(note: NoteEntity)
        fun onNoteSelectedForEditing(note: NoteEntity, appendMode: Boolean)
    }

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<NoteEntity>() {
            override fun areItemsTheSame(oldItem: NoteEntity, newItem: NoteEntity): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: NoteEntity, newItem: NoteEntity): Boolean {
                return oldItem == newItem
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_note, parent, false)
        return NoteViewHolder(view)
    }

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class NoteViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val noteText: TextView = itemView.findViewById(R.id.noteText)
        private val starIcon: ImageView = itemView.findViewById(R.id.starIcon)

        fun bind(note: NoteEntity) {
            noteText.text = note.content
            starIcon.setImageResource(
                if (note.isFavorite) R.drawable.ic_star_filled
                else R.drawable.ic_star_outline
            )

            starIcon.setOnClickListener {
                note.isFavorite = !note.isFavorite
                starIcon.setImageResource(
                    if (note.isFavorite) R.drawable.ic_star_filled
                    else R.drawable.ic_star_outline
                )
                listener.onNoteChanged(note)
            }

            itemView.setOnClickListener {
                listener.onNoteSelectedForEditing(note, appendMode = false)
            }

            itemView.setOnLongClickListener {
                listener.onNoteSelectedForEditing(note, appendMode = true)
                true
            }
        }
    }
}
